### PR TITLE
Ping, Pong: make immutable

### PR DESCRIPTION
**Pull request by schildbach**

### CODE REVIEW

**Comment by schildbach** [comment]
This is an example of how I would like the Message classes look like.

Note that Ping and Pong so simple they will probably never have data classes (which would go to base). Their "data class" is a long, for the nonce.

Note that the classes are entirely immutable because they extend the still mutable Message class. But at least the classes don't reference the mutable fields (serializer and payload).

**Review by msgilligan — CHANGES_REQUESTED** [review]
Overall, this looks great! This is a good sign!

I like the `.read(ByteBuffer)` factory method.

I have some minor request and questions, though.

**Comment by msgilligan** [comment]
> The classes are **not** entirely immutable because they extend the still mutable Message class. But at least the classes don't reference the mutable fields (`serializer` and `payload`).

I fixed what I assume was a typo above by adding the **not**.

File: core/src/main/java/org/bitcoinj/core/Ping.java
Comment (schildbach) [reply core/src/main/java/org/bitcoinj/core/Ping.java:69]: Right, but it's just a private constructor. I'm kind of okay with the permutation explosion for public APIs, but for an internal method?

File: core/src/main/java/org/bitcoinj/core/Ping.java
Comment (schildbach) [reply core/src/main/java/org/bitcoinj/core/Ping.java:34]: Yes, to discourage nonce-less pings.

File: core/src/main/java/org/bitcoinj/core/Peer.java
Comment (schildbach) [reply core/src/main/java/org/bitcoinj/core/Peer.java:1604]: Hmmmm… :confused:

**Comment by schildbach** [comment]
But actually more than the implementation details I wanted to present an example of how the future could look like.

On that basis, I'd like to start a discussion on how more complex cases like Transactions would look like.

File: core/src/main/java/org/bitcoinj/core/Peer.java
Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/core/Peer.java:1604]: Perhaps with line-breaks:
```
m.nonce()
    .map(Pong::of)
    .ifPresent(this::sendMessage);
```

File: core/src/main/java/org/bitcoinj/core/Ping.java
Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/core/Ping.java:69]: I think its worth it to eliminate a boxed `Long` and nullability (not to mention the `@Nullabble` annotation.

**Comment by msgilligan** [comment]
> We could also remove the nonceless Ping case … nobody is running such old nodes any more.

That is worth considering, but should probably be a separate Issue/PR.

**Comment by msgilligan** [comment]
> But actually more than the implementation details I wanted to present an example of how the future could look like.

I definitely give this a "concept ack".

> On that basis, I'd like to start a discussion on how more complex cases like Transactions would look like.

What's the next step in that discussion?  I guess I should look at what's left in `Message` that is mutable, for starters...

File: core/src/main/java/org/bitcoinj/core/Ping.java
Comment (msgilligan) [reply core/src/main/java/org/bitcoinj/core/Ping.java:69]: Alternatively, it could be `Ping(long nonce, boolean hasNonce)`

**Comment by schildbach** [comment]
I pushed an update and I think the two messages are now 100% immutable.

**Comment by schildbach** [comment]
> > We could also remove the nonceless Ping case … nobody is running such old nodes any more.
> 
> That is worth considering, but should probably be a separate Issue/PR.

It would be extremely easy to implement in current master, before we merge this.

**Comment by schildbach** [comment]
The current behavior (and the one re-implemented in this PR) is not accurate anyhow. According to BIP31, protocol versions from 60000 upwards have to use (and expect) a nonce.

https://github.com/bitcoin/bips/blob/master/bip-0031.mediawiki

**Comment by schildbach** [comment]
See https://github.com/bitcoinj/bitcoinj/pull/2861.

**Comment by schildbach** [comment]
I rebased this draft on all our recent work.

Because we decided to remove support for nonce-less pings a few days ago, I now dropped `Optional` and introduced a new helper: `Ping.pong()`.

File: core/src/main/java/org/bitcoinj/core/Peer.java
Comment (schildbach) [reply core/src/main/java/org/bitcoinj/core/Peer.java]: We got rid of the `Optional`, so I resolve this.

File: core/src/main/java/org/bitcoinj/core/Ping.java
Comment (schildbach) [reply core/src/main/java/org/bitcoinj/core/Ping.java:67]: Nonces are now required, so I resolve this.

File: core/src/main/java/org/bitcoinj/core/Ping.java
Comment (schildbach) [reply core/src/main/java/org/bitcoinj/core/Ping.java]: We got rid of the `Optional`, so I resolve this.

File: core/src/main/java/org/bitcoinj/core/Ping.java
Comment (schildbach) [reply core/src/main/java/org/bitcoinj/core/Ping.java]: We got rid of the `Optional`, so I resolve this.

File: integration-test/src/test/java/org/bitcoinj/core/FilteredBlockAndPartialMerkleTreeTest.java
Comment (schildbach) [reply integration-test/src/test/java/org/bitcoinj/core/FilteredBlockAndPartialMerkleTreeTest.java]: There is now a `Pong pong = ping.pong()`.

**Review by msgilligan — APPROVED** [review]
Looks great! Why is this still draft?

**Comment by schildbach** [comment]
I'm still extracting refactorings for it. For now, it's mainly meant as a vision.

**Comment by schildbach** [comment]
Next step is `parse()` → `static read()`.

**Comment by msgilligan** [comment]
> Next step is `parse()` → `static read()`.

If PR #3006 is merged and this PR is rebased it can be merged without making a the global  `parse()` → `static read()` change.

**Comment by schildbach** [comment]
Yes, this could be merged any time, even without rebasing. I'm holding off because I want to make sure we manage to apply the individual refactoring steps on the entire hierarchy. We've had too many migrations stuck at some half-way point in the past, so I'm a bit reluctant.

Currently we're stuck at `Transaction`, and the critical blocker is the `parent` issue. But at least a couple of more sub-messages got divorced from `Message`, which is also great progress imho.

Let me know if the merge status of this PR is blocking anything you're working on.

**Comment by schildbach** [comment]
Merged.

**Pull request closed without merging**
